File: iOS/Sources/VertuEdge/RPA/FlowParser.swift
import Foundation

struct FlowParser {
    func parse(_ yaml: String) throws -> FlowDefinition {
        let parts = yaml.components(separatedBy: "\n---\n")
        let header = parts.first ?? ""
        let body = parts.count > 1 ? parts[1] : ""

        guard let appId = extractValue(from: header, key: "appId") else {
            throw FlowParseError.missingAppId
        }
        let name = extractValue(from: header, key: "name")
        let actions = parseActions(from: body)
        return FlowDefinition(appId: appId, name: name, actions: actions)
    }

    private func extractValue(from text: String, key: String) -> String? {
        let pattern = "^\(key):\\s*(.+)$"
        guard let regex = try? NSRegularExpression(pattern: pattern, options: .anchorsMatchLines),
              let match = regex.firstMatch(in: text, range: NSRange(text.startIndex..., in: text)),
              let range = Range(match.range(at: 1), in: text) else { return nil }
        return String(text[range]).trimmingCharacters(in: .whitespaces)
            .trimmingCharacters(in: CharacterSet(charactersIn: "\""))
    }

    private func parseActions(from body: String) -> [FlowAction] {
        var actions: [FlowAction] = []
        for line in body.components(separatedBy: "\n") {
            let t = line.trimmingCharacters(in: .whitespaces)
            if t.hasPrefix("- launchApp:") {
                actions.append(.launchApp(appId: value(t, prefix: "- launchApp:")))
            } else if t.hasPrefix("- tapOn:") {
                actions.append(.tapOn(selector: value(t, prefix: "- tapOn:")))
            } else if t.hasPrefix("- inputText:") {
                actions.append(.inputText(value(t, prefix: "- inputText:")))
            } else if t.hasPrefix("- assertVisible:") {
                actions.append(.assertVisible(value(t, prefix: "- assertVisible:")))
            } else if t.hasPrefix("- assertNotVisible:") {
                actions.append(.assertNotVisible(value(t, prefix: "- assertNotVisible:")))
            } else if t.hasPrefix("- scrollUntilVisible:") {
                actions.append(.scrollUntilVisible(selector: value(t, prefix: "- scrollUntilVisible:")))
            } else if t.hasPrefix("- openLink:") {
                actions.append(.openLink(url: value(t, prefix: "- openLink:")))
            } else if t.hasPrefix("- wait:") {
                let ms = Int(value(t, prefix: "- wait:")) ?? 1000
                actions.append(.wait(durationMs: ms))
            } else if t.hasPrefix("- pressKey:") {
                actions.append(.pressKey(value(t, prefix: "- pressKey:")))
            } else if t.hasPrefix("- runAiPrompt:") {
                actions.append(.runAiPrompt(prompt: value(t, prefix: "- runAiPrompt:")))
            } else if t.hasPrefix("- takeScreenshot") {
                let label = t.contains(":") ? value(t, prefix: "- takeScreenshot:") : nil
                actions.append(.takeScreenshot(label: label?.isEmpty == true ? nil : label))
            } else if t.hasPrefix("- clearState:") {
                actions.append(.clearState(appId: value(t, prefix: "- clearState:")))
            }
        }
        return actions
    }

    private func value(_ line: String, prefix: String) -> String {
        line.dropFirst(prefix.count)
            .trimmingCharacters(in: .whitespaces)
            .trimmingCharacters(in: CharacterSet(charactersIn: "\""))
    }
}

enum FlowParseError: Error {
    case missingAppId
    case invalidYaml(String)
}

File: Android/app/src/main/java/com/vertu/edge/rpa/FlowParser.kt
package com.vertu.edge.rpa

import com.vertu.edge.common.AppConstants

data class FlowDefinition(
    val appId: String,
    val name: String? = null,
    val actions: List<FlowAction> = emptyList()
)

class FlowParser {

    fun parse(yaml: String): Result<FlowDefinition> = runCatching {
        val lines = yaml.lines()
        val headerLines = mutableListOf<String>()
        val actionLines = mutableListOf<String>()
        var foundSeparator = false

        for (line in lines) {
            if (line.trim() == "---") {
                foundSeparator = true
                continue
            }
            if (foundSeparator) actionLines.add(line) else headerLines.add(line)
        }

        val header = headerLines.joinToString("\n")
        val appId = extractValue(header, "appId") ?: error("appId is required in flow header")
        val name = extractValue(header, "name")

        val actions = parseActions(actionLines)
        FlowDefinition(appId = appId, name = name, actions = actions)
    }

    private fun extractValue(text: String, key: String): String? {
        val regex = Regex("^$key:\\s*(.+)$", RegexOption.MULTILINE)
        return regex.find(text)?.groupValues?.getOrNull(1)?.trim()?.removeSurrounding("\"")
    }

    private fun parseActions(lines: List<String>): List<FlowAction> {
        val actions = mutableListOf<FlowAction>()
        var i = 0
        while (i < lines.size) {
            val line = lines[i].trim()
            when {
                line.startsWith("- launchApp:") -> {
                    val appId = line.removePrefix("- launchApp:").trim().removeSurrounding("\"")
                    actions.add(FlowAction.LaunchApp(appId))
                }
                line == "- launchApp" -> {
                    actions.add(FlowAction.LaunchApp(""))
                }
                line.startsWith("- tapOn:") -> {
                    val selector = line.removePrefix("- tapOn:").trim().removeSurrounding("\"")
                    actions.add(FlowAction.TapOn(selector))
                }
                line.startsWith("- inputText:") -> {
                    val text = line.removePrefix("- inputText:").trim().removeSurrounding("\"")
                    actions.add(FlowAction.InputText(text))
                }
                line.startsWith("- assertVisible:") -> {
                    val selector = line.removePrefix("- assertVisible:").trim().removeSurrounding("\"")
                    actions.add(FlowAction.AssertVisible(selector))
                }
                line.startsWith("- assertNotVisible:") -> {
                    val selector = line.removePrefix("- assertNotVisible:").trim().removeSurrounding("\"")
                    actions.add(FlowAction.AssertNotVisible(selector))
                }
                line.startsWith("- scrollUntilVisible:") -> {
                    val selector = line.removePrefix("- scrollUntilVisible:").trim().removeSurrounding("\"")
                    actions.add(FlowAction.ScrollUntilVisible(selector))
                }
                line.startsWith("- openLink:") -> {
                    val url = line.removePrefix("- openLink:").trim().removeSurrounding("\"")
                    actions.add(FlowAction.OpenLink(url))
                }
                line.startsWith("- wait:") -> {
                    val ms = line.removePrefix("- wait:").trim().toLongOrNull()
                        ?: AppConstants.DEFAULT_FLOW_TIMEOUT_MS
                    actions.add(FlowAction.Wait(ms))
                }
                line.startsWith("- pressKey:") -> {
                    val key = line.removePrefix("- pressKey:").trim().removeSurrounding("\"")
                    actions.add(FlowAction.PressKey(key))
                }
                line.startsWith("- runAiPrompt:") -> {
                    val prompt = line.removePrefix("- runAiPrompt:").trim().removeSurrounding("\"")
                    actions.add(FlowAction.RunAiPrompt(prompt))
                }
                line.startsWith("- takeScreenshot") -> {
                    val label = if (line.contains(":")) {
                        line.substringAfter(":").trim().removeSurrounding("\"").ifEmpty { null }
                    } else null
                    actions.add(FlowAction.TakeScreenshot(label))
                }
                line.startsWith("- clearState:") -> {
                    val appId = line.removePrefix("- clearState:").trim().removeSurrounding("\"")
                    actions.add(FlowAction.ClearState(appId))
                }
            }
            i++
        }
        return actions
    }
}
